contracts/exchange/Exchange_ERC20_1155.cairo

```diff
@@ -0,0 +1,646 @@
+    end
+
+    let (caller) = get_caller_address()
+    let (contract) = get_contract_address()
+
+    let (token_address_) = token_address.read()
+    let (currency_address_) = currency_address.read()
+
+    # Read current reserve levels
+    let (currency_reserves_) = currency_reserves.read(token_id)
+    let (token_reserves) = IERC1155.balanceOf(token_address_, contract, token_id)
+
+    # Transfer max currency form caller to contract
+    IERC20.transferFrom(currency_address_, caller, contract, max_currency_amount)
+    tempvar syscall_ptr :felt* = syscall_ptr
+
+    # Calculate prices
+    let (currency_amount) = get_buy_price(token_amount, currency_reserves_, Uint256(token_reserves, 0))
+
+    # Calculate refund
+    let (refund_amount) = uint256_sub(max_currency_amount, currency_amount)
+
+    # Update reserves
+    let (new_reserves, add_overflow) = uint256_add(currency_reserves_, currency_amount)
+    with_attr error_message("Currency value overflow"):
+        assert add_overflow = 0
+    end
+    currency_reserves.write(token_id, new_reserves)
+
+    # Transfer refunded currency and purchased tokens
+    IERC20.transfer(currency_address_, caller, refund_amount)
+    tempvar syscall_ptr :felt* = syscall_ptr
+    IERC1155.safeTransferFrom(token_address_, contract, caller, token_id, token_amount.low)
+
+    # Emit event
+    tokens_purchased.emit(caller, currency_amount, token_id, token_amount)
+
+    return (currency_amount)
+end
+
+
+#FIXME IERC1155 doesn't have a call back when using ERC1155.safeTransfer.
+# User will need to `setApprovalForAll` to this contract, which poses a security risk for repeat transactions.
+@external
+func sell_tokens {
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr,
+    }(
+        # Maximum amount of currency to buy
+        min_currency_amount: Uint256,
+        # ERC1155 token id
+        token_id: felt,
+        # Exact amount of token to sell
+        token_amount: Uint256,
+        # Maximum time which this transaction can be accepted
+        deadline: felt,
+    ) -> (
+        sold: Uint256
+    ):
+    alloc_locals
+
+    # Check deadline within bounds
+    let (block_timestamp) = get_block_timestamp()
+    with_attr error_message("Deadline exceeded"):
+        assert_le(block_timestamp, deadline)
+    end
+
+    let (caller) = get_caller_address()
+    let (contract) = get_contract_address()
+
+    let (token_address_) = token_address.read()
+    let (currency_address_) = currency_address.read()
+
+    # Read current reserve levels
+    let (currency_reserves_) = currency_reserves.read(token_id)
+    let (token_reserves) = IERC1155.balanceOf(token_address_, contract, token_id)
+
+    # Take the token amount
+    IERC1155.safeTransferFrom(token_address_, caller, contract, token_id, token_amount.low)
+
+    # Calculate prices
+    let (currency_amount) = get_sell_price(token_amount, currency_reserves_, Uint256(token_reserves, 0))
```

Comment:
What [Nifty does here](https://github.com/0xsequence/niftyswap/blob/master/src/contracts/exchange/NiftyswapExchange20.sol#L254-L257) is to subtract the amount of tokens sold from the token reserves (not their argument positions are different to the `get_sell_price` in this  contract).

Reply:
Yes, that's because they transfer the tokens in before reading the reserve amount. We are reading the balance and then transferring the tokens.